File: backend/app/schemas/audit_log.py
"""
Pydantic schemas for audit logs.
"""

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict, Field


class LogResourceBase(BaseModel):
    schema_name: str = Field(..., description="Schema name")
    table_name: str = Field(..., description="Table name")
    record_id: Optional[str] = Field(None, description="Primary key as string")
    record_pk: Optional[dict] = Field(None, description="Composite primary key")
    biz_key: Optional[str] = Field(None, description="Business key")
    op: str = Field(..., description="Operation code")
    before: Optional[Any] = None
    after: Optional[Any] = None
    diff: Optional[Any] = None
    sensitivity_level: int = Field(0, ge=0, le=2)


class LogResourceCreate(LogResourceBase):
    pass


class LogResourceOut(LogResourceBase):
    id: int
    log_id: int
    created_at: Optional[datetime] = None

    model_config = ConfigDict(
        from_attributes=True,
    )


class LogEntryBase(BaseModel):
    ts: Optional[datetime] = None
    user_id: Optional[int] = None
    username: Optional[str] = None
    real_name: Optional[str] = None
    action: str
    action_display: Optional[str] = None
    action_category: Optional[str] = None
    module: Optional[str] = None
    endpoint: Optional[str] = None
    method: Optional[str] = None
    status_code: Optional[int] = None
    latency_ms: Optional[int] = None
    ip: Optional[str] = None
    user_agent: Optional[str] = None
    request_id: Optional[str] = None
    success: bool = True
    error_message: Optional[str] = None
    resource_summary: Optional[Any] = None
    extra: Optional[Any] = None


class LogEntryCreate(LogEntryBase):
    resources: list[LogResourceCreate] = Field(default_factory=list)


class LogEntryOut(LogEntryBase):
    id: int
    resources: Optional[list[LogResourceOut]] = None

    model_config = ConfigDict(
        from_attributes=True,
    )


class LogListResponse(BaseModel):
    total: int
    items: list[LogEntryOut]

File: frontend/services/auditLog.ts
// [Common Module] Audit log service
import { api } from './api'

export interface AuditLogResource {
  id: number
  log_id: number
  schema_name: string
  table_name: string
  record_id?: string
  record_pk?: Record<string, any>
  biz_key?: string
  op: string
  before?: any
  after?: any
  diff?: any
  sensitivity_level: number
  created_at?: string
}

export interface AuditLogEntry {
  id: number
  ts?: string
  user_id?: number
  username?: string
  real_name?: string
  action: string
  action_display?: string
  action_category?: string
  module?: string
  endpoint?: string
  method?: string
  status_code?: number
  latency_ms?: number
  ip?: string
  user_agent?: string
  request_id?: string
  success: boolean
  error_message?: string
  resource_summary?: any
  extra?: any
  resources?: AuditLogResource[]
}

export interface AuditLogListResponse {
  total: number
  items: AuditLogEntry[]
}

export const listAuditLogs = async (params: {
  user_id?: number
  username?: string
  action?: string
  action_category?: string
  module?: string
  table_name?: string
  record_id?: string
  start_time?: string
  end_time?: string
  success?: boolean
  limit?: number
  offset?: number
  include_resources?: boolean
}) => {
  const res = await api.get('/logs', { params })
  return res.data as AuditLogListResponse
}

export const getAuditLog = async (logId: number, includeResources = true) => {
  const res = await api.get(`/logs/${logId}`, {
    params: { include_resources: includeResources },
  })
  return res.data as AuditLogEntry
}
